source/declarations.tex

```diff
@@ -2868,159 +2868,132 @@
-is
-``\placeholder{derived-declarator-type-list}
-\tcode{T}'',
-then the type of the identifier of
-\tcode{D}
-is an array type; if the type of the identifier of \tcode{D}
-contains the \tcode{auto} \grammarterm{type-specifier},
-the program is ill-formed.
-\tcode{T}
-is called the array
-\term{element type};
-this type shall not be a reference type, \cv{}~\tcode{void},
-or a function type.
-\indextext{declaration!array}%
-If the
-\grammarterm{constant-expression}\iref{expr.const}
-is present, it shall be a converted constant
-expression of type \tcode{std::size_t} and
-its value shall be greater than zero.
-The constant expression specifies the
-\indextext{array!bound}%
+and the type of the contained \grammarterm{declarator-id}
+in the declaration \tcode{T} \tcode{D1}
+is ``\placeholder{derived-declarator-type-list} \tcode{T}'',
+the type of the \grammarterm{declarator-id} in \tcode{D} is
+``\placeholder{derived-declarator-type-list} array of \tcode{N} \tcode{T}''.
+The \grammarterm{constant-expression}
+shall be a converted constant expression of type \tcode{std::size_t}\iref{expr.const}.
 \indextext{bound, of array}%
-\term{bound}
-of (number of elements in) the array.
-If the value of the constant expression is
-\tcode{N},
-the array has
-\tcode{N}
-elements numbered
-\tcode{0}
-to
-\tcode{N-1},
-and the type of the identifier of
-\tcode{D}
-is ``\placeholder{derived-declarator-type-list} array of
-\tcode{N}
-\tcode{T}''.
-An object of array type contains a contiguously allocated non-empty set of
-\tcode{N}
-subobjects of type
-\tcode{T}.
-Except as noted below, if
-the constant expression is omitted, the type of the identifier of
-\tcode{D}
-is ``\placeholder{derived-declarator-type-list} array of unknown bound of
-\tcode{T}'',
-an incomplete object type.
-The type ``\placeholder{derived-declarator-type-list} array of
-\tcode{N}
-\tcode{T}''
-is a different type from the type
-``\placeholder{derived-declarator-type-list} array of unknown bound of
-\tcode{T}'',
-see~\ref{basic.types}.
+Its value \tcode{N} specifies the \defnx{array bound}{array!bound},
+i.e., the number of elements in the array;
+\tcode{N} shall be greater than zero.
+An object of such a type contains
+a contiguously allocated non-empty set
+of \tcode{N} subobjects of type \tcode{T},
+known as the \defnx{elements}{array!element} of the array,
+and numbered \tcode{0} to \tcode{N-1}.
+
+\pnum
+In a declaration \tcode{T} \tcode{D} where \tcode{D} has the form
+
+\begin{ncsimplebnf}
+\terminal{D1 [ ]} \opt{attribute-specifier-seq}
+\end{ncsimplebnf}
+
+and the type of the contained \grammarterm{declarator-id}
+in the declaration \tcode{T} \tcode{D1}
+is ``\placeholder{derived-declarator-type-list} \tcode{T}'',
+the type of the \grammarterm{declarator-id} in \tcode{D} is
+``\placeholder{derived-declarator-type-list} array of unknown bound of \tcode{T}'', except as specified below.
+
+\pnum
+\indextext{declaration!array}%
+In either case, the type is an \defn{array type}.
```

Comment:
This seems untrue. Given `int (*p)[3];`, we apply the above rule with `T` = `int`, `D` = `(*p)[3]`, and `D1` = `(*p)`, and we're correctly told that the type of `p` is "pointer to array of 3 int". Then this paragraph says that that type is an array type, which it isn't.

This bug is pre-existing, but I thought this rework was supposed to fix it. Maybe we should say that an array type is a type of the form "array of N T" or of the form "array of unknown bound of T", and the element type is T, here.

Reply:
Fixed.